Sources/Profile-Modify.php

```diff
@@ -2342,110 +2347,130 @@ function alert_count($memID, $unread = false)
+		}
+		elseif ($row['content_type'] == 'unapproved' && $row['content_action'] == 'attachment')
+		{
+			$alerts[$row['id_alert']]['visible'] = false;
+			$possible_attachments[$row['id_alert']] = $row['content_id'];
 		}
-		elseif (isset($alerts[$row['id_alert']]['topic']))
-			$possible_topics[] = $alerts[$row['id_alert']]['topic'];
-		elseif (isset($alerts[$row['id_alert']]['board']))
-			$possible_boards[] = $alerts[$row['id_alert']]['board'];
+		// For the rest, they can always see it.
+		else
+			$alerts[$row['id_alert']]['visible'] = true;
 	}
 	$smcFunc['db_free_result']($request);
 
-	$possible_boards = array_unique($possible_boards);
-
-	// If this isn't the current user, get their boards.
-	if (!isset($user_info['id']) || $user_info['id'] != $memID)
+	// If we need to check board access, use the correct board access filter for the member in question.
+	if ((!isset($user_info) || $user_info['id'] != $memID) && (!empty($possible_msgs) || !empty($possible_topics) || !empty($possible_attachments)))
+		$qb = build_query_board($memID);
+	else
 	{
-		$query_see_board = build_query_board($memID);
-		$query_see_board = $query_see_board['query_see_board'];
+		$qb['query_see_topic_board'] = '{query_see_topic_board}';
+		$qb['query_see_message_board'] = '{query_see_message_board}';
 	}
-	else
-		$query_see_board = '{query_see_board}';
 
 	// We want only the stuff they can see.
 	if (!empty($possible_msgs))
 	{
+		$flipped_msgs = array();
+		foreach ($possible_msgs as $id_alert => $id_msg)
+		{
+			if (!isset($flipped_msgs[$id_msg]))
+				$flipped_msgs[$id_msg] = array();
+
+			$flipped_msgs[$id_msg][] = $id_alert;
+		}
+
 		$request = $smcFunc['db_query']('', '
-			SELECT m.id_msg, m.id_topic, b.id_board
+			SELECT m.id_msg
 			FROM {db_prefix}messages AS m
-				INNER JOIN {db_prefix}boards AS b ON (m.id_board = b.id_board)
-			WHERE ' . $query_see_board . '
+			WHERE ' . $qb['query_see_message_board'] . '
 				AND m.id_msg IN ({array_int:msgs})',
 			array(
 				'msgs' => $possible_msgs,
 			)
 		);
 		while ($row = $smcFunc['db_fetch_assoc']($request))
 		{
-			$msgs[] = $row['id_msg'];
-			$topics[] = $row['id_topic'];
-			$boards[] = $row['id_board'];
+			foreach ($flipped_msgs[$row['id_msg']] as $id_alert)
+				$alerts[$id_alert]['visible'] = true;
 		}
-
 		$smcFunc['db_free_result']($request);
 	}
 	if (!empty($possible_topics))
 	{
+		$flipped_topics = array();
+		foreach ($possible_topics as $id_alert => $id_topic)
+		{
+			if (!isset($flipped_topics[$id_topic]))
+				$flipped_topics[$id_topic] = array();
+
+			$flipped_topics[$id_topic][] = $id_alert;
+		}
+
 		$request = $smcFunc['db_query']('', '
-			SELECT t.id_topic, b.id_board
+			SELECT t.id_topic
 			FROM {db_prefix}topics AS t
-				INNER JOIN {db_prefix}boards AS b ON (t.id_board = b.id_board)
-			WHERE ' . $query_see_board . '
+			WHERE ' . $qb['query_see_topic_board'] . '
 				AND t.id_topic IN ({array_int:topics})',
 			array(
 				'topics' => $possible_topics,
```

Comment:
I am getting `Invalid value inserted or no type specified.` on this line when locking/stickying a topic

Reply:
Can you give me the backtrace info for that error and a bit more info on how to reproduce it? Was that using quick moderation or the buttons in the topic display? What were the permissions of the user that triggered the error?

Reply:
```
#0: smf_db_error_backtrace()
Called from ...\Sources\Subs-Db-mysql.php on line 190
#1: smf_db_replacement__callback()
Called from unknown on line -1
#2: preg_replace_callback()
Called from ...\Sources\Subs-Db-mysql.php on line 404
#3: smf_db_query()
Called from ...\Sources\Profile-Modify.php on line 2431
#4: alert_count()
Called from ...\Sources\Subs.php on line 398
#5: updateMemberData()
Called from ...\Sources\tasks\CreatePost-Notify.php on line 358
#6: execute()
Called from ...\cron.php on line 216
#7: perform_task()
Called from ...\cron.php on line 110
```

user is an admin, the action was taken from the topic page

Reply:
Derp. I was using `{query_see_message_board` and` {query_see_topic_board}`, but those were added in a different branch. 82d8de5 should fix it.